brave/brave-core
# Reworked identity patch for auth token

Resolves https://github.com/brave/brave-browser/issues/29096

In last upstream Google has started to block https://source.chromium.org/chromium/chromium/src/+/b4c9aeeaa5b3cfa59d6e9f1bb0d0d012f5f77ec4 getAuthToken interactive calls when idle. This breaks the old patch because
the fallback to WebAuthFlow can only be completed in interactive mode https://chromium-review.googlesource.com/c/chromium/src/+/2704571/4/chrome/browser/extensions/api/identity/identity_get_auth_token_function.cc#526.

From docs about Interactive versus silent mode (
https://developer.chrome.com/docs/apps/app_identity/#launchWebAuthFlow-interactive)

`If you invoke the API in silent mode, the API will only return a token if the provider is able to provide a token without showing any UI. This is useful in cases when an app is doing the flow at app startup, for example, or in general in cases where there is no user gesture involved.`

It means that silent mode should be supported by server side and return a token.

Samples extensions used for tests to authorize on Google servers were not authorized in silent mode and working only in interactive.

In order to fix this I force the interactive mode in fallback to WebAuthFlow each time when it is possible or explicitly requested by api, otherwise the silent mode will be used if no silent mode allowed AND it was request by api call.

It looks valid because the sense of initial chromium patch is to redirect silent modes to interactive using WebAuthFlow https://crrev.com/c/2704571



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

- Install https://chrome.google.com/webstore/detail/identity-api-sample/oficlfehfenioickohognhdhmmcpceil and try to login
- check token received and data can be fetched
